lib/HLSL/DxilSimpleGVNHoist.cpp

```diff
@@ -585,3 +587,180 @@ FunctionPass *llvm::createDxilSimpleGVNHoistPass() {
+  unsigned InstIndex = 0;
+  unsigned TerminatorIndex = 0;
+
+  struct TerminatorRecord {
+    uint32_t Opcode;
+    uint32_t ConditionVN;
+  };
+  std::vector<uint32_t> InstValueNumbers;
+  std::vector<TerminatorRecord> TerminatorRecords;
+
+  bool HandleInst(ValueTable &VT, Instruction *I, bool Cmp) {
+    if (isa<DbgInfoIntrinsic>(I))
+      return true;
+    if (TerminatorInst *TI = dyn_cast<TerminatorInst>(I)) {
+      Value *Cond = nullptr;
+      if (BranchInst *Br = dyn_cast<BranchInst>(TI)) {
+        if (!Br->isConditional())
+          return true;
+        Cond = Br->getCondition();
+      }
+      else if (SwitchInst *Sw = dyn_cast<SwitchInst>(TI)) {
+        Cond = Sw->getCondition();
+      }
+      // Don't know how to handle this. Don't assume anything.
+      else {
+        return false;
+      }
+
+      // Either add a record or compare it to the record
+      if (!Cmp) {
+        TerminatorRecord NewRec = {};
+        NewRec.ConditionVN = VT.lookupOrAdd(Cond);
+        NewRec.Opcode = TI->getOpcode();
+        TerminatorRecords.push_back(NewRec);
+      }
+      else {
+        if (TerminatorIndex >= TerminatorRecords.size())
+          return false;
+
+        TerminatorRecord Record = TerminatorRecords[TerminatorIndex];
+        if (Record.Opcode != TI->getOpcode() || Record.ConditionVN != VT.lookup(Cond, false))
+          return false;
+
+        TerminatorIndex++;
+      }
+    }
+    else {
+      // Either add a record or compare it to the record
+      if (!Cmp) {
+        uint32_t VN = VT.lookupOrAdd(I);
+        InstValueNumbers.push_back(VN);
+      }
+      else {
+        uint32_t VN = VT.lookupOrAdd(I);
+        if (InstIndex >= InstValueNumbers.size() || InstValueNumbers[InstIndex] != VN)
+          return false;
+        InstIndex++;
+      }
+    }
+
+    return true;
+  }
+
+  bool BuildOrCompare(BasicBlock *Begin, BasicBlock *End, ValueTable &VT, bool Compare) {
+    SmallVector<BasicBlock *, 10> Worklist;
+    DenseSet<BasicBlock *> Seen;
+    Worklist.push_back(Begin);
+
+    while (Worklist.size()) {
+      BasicBlock *BB = Worklist.pop_back_val();
+      if (BB == End)
+        continue;
+      Seen.insert(BB);
+
+      for (Instruction &I : *Begin) {
+        if (!HandleInst(VT, &I, Compare))
+          return false;
+      }
+
+      for (BasicBlock *Succ : successors(BB)) {
+        // If the successor was encountered before, there might be a loop in here somewhere.
+        // Don't risk it.
+        if (Seen.count(Succ)) {
+          return false;
+        }
+        Worklist.push_back(Succ);
+      }
+    }
+
+    if (!Compare) {
+      return true;
+    }
+    else {
+      return InstValueNumbers.size() == InstIndex && TerminatorRecords.size() == TerminatorIndex;
+    }
+  }
+
+};
+
+bool DxilSimpleGVNEliminateRegion::runOnFunction(Function &F) {
+  PostDominatorTree *PDT = &getAnalysis<PostDominatorTree>();
+
+  ValueTable VT;
+
+  bool bChanged = false;
+  for (BasicBlock &BB : F) {
```

Comment:
I'm worried about the time complexity of this whole walk. It looks like an N^2 walk since we can visit a block multiple times (so if we have highly nested ifs we are going to be vising the same block many times).

It would be good to think of a way to structure it so that we only visit each block once.

Reply:
In this new version, the expensive thing, which is to go through every instruction, is limited to running only once (twice if a block is modified), so it's N at this point.

I don't know if it' actually worth it to try to not re-walk blocks when looking for regions.